Add InList and NotInList

fahimbagar opened the pull request:
This PR does the following:
- Add `InList` and `NotInList` for given types:
  - [x] Value
  - [x] Object
  - [x] Array
  - [x] String
  - [x] Number
  - [x] Boolean (just for completeness)
  - [x] Duration
  - [x] DateTime
-Since `AssertionList` rejects a single slice as a param, single array param in `InList` will behave like `IsEqual` and `NotInList` will behave like `NotEqual`. Addressed in commit https://github.com/gavv/httpexpect/commit/649ee8caf0e8c918aa9a2de8a8d6f6c5c3ec1cf7

https://github.com/gavv/httpexpect/blob/86f366a2d10b8a8ddd3cb1a9b2a5c00f37a1df68/assertion_validation.go#L186-L190

For issue https://github.com/gavv/httpexpect/issues/250

gavv commented:
I think we should have separate check for zero-length list.

Such call will always fail, but AssertBelongs will look confusing.

Since it anyway will fail, let's forbid this case. If length of list is zero, let's fail with AssertUsage with error "unexpected empty list".

fahimbagar commented:
> Since it anyway will fail, let's forbid this case. If length of list is zero, let's fail with AssertUsage with error "unexpected empty list argument".

> Also would be nice to cover these usage checks with tests.

Addressed both, let me know if I miss out anything. Also removed unneeded `canonArray` in https://github.com/gavv/httpexpect/pull/256/commits/b7fc213ce722a152f674015a898290fe3475e965 and https://github.com/gavv/httpexpect/pull/256/commits/b885435c99ef42ba2d7c82b34021ef29f6de229c

gavv commented on array.go:
```diff
@@ -746,6 +746,128 @@ func (a *Array) Equal(value interface{}) *Array {
 	return a.IsEqual(value)
 }
 
+// InList succeeds if array is listed by given [values...].
```
Here and in other methods, I suggest to reword this part a bit:

```
InList succeeds if whole array is equal to one of the elements from given list of arrays.
```

fahimbagar replied:
Addressed in `Array` and others as well.

gavv commented on array.go:
```diff
@@ -746,6 +746,128 @@ func (a *Array) Equal(value interface{}) *Array {
 }
 
+// InList succeeds if array is listed by given [values...].
+// Before comparison, both array and value are converted to canonical form.
```
Let's reword this a bit too:

```
Before comparison, both array and each value are converted to canonical form.
```

fahimbagar replied:
Addressed in `Array` and others as well.

gavv commented on array.go:
```diff
@@ -746,6 +746,128 @@ func (a *Array) Equal(value interface{}) *Array {
+// InList succeeds if array is listed by given [values...].
+// Before comparison, both array and value are converted to canonical form.
+//
+// values should be an array of slice of any type. This comparison adheres
+// element order.
```
And this:

```
Each value should be a slice of any type.
```

fahimbagar replied:
Addressed in `Array` and `Number`

gavv commented on array.go:
```diff
@@ -746,6 +746,128 @@ func (a *Array) Equal(value interface{}) *Array {
+// values should be an array of slice of any type. This comparison adheres
+// element order.
+//
+// Example:
+//
+//	array := NewArray(t, []interface{}{"foo", 123})
+//	array.InList([]interface{}{"foo", 123})
+//
+//	array := NewArray(t, []interface{}{"foo", "bar"})
+//	array.InList([]string{}{"foo", "bar"})
+//
+//	array := NewArray(t, []interface{}{123, 456})
+//	array.InList([]int{}{123, 456})
```
Let's provide example with multiple arguments to InList, to make it clear that it accepts list of slices, not a single slice.

fahimbagar replied:
Addressed in `Array` and others as well.

gavv commented on array.go:
```diff
@@ -746,6 +746,128 @@ func (a *Array) Equal(value interface{}) *Array {
+			Actual:   &AssertionValue{a.value},
+			Expected: &AssertionValue{AssertionList(values)},
+			Errors: []error{
+				errors.New("expected: arrays are listed"),
```
Here and in other methods, let's reword error message:

```
expected: arrays is equal to one of the values
```

fahimbagar replied:
Addressed in `Array` and others as well.

gavv commented on boolean.go:
```diff
@@ -162,6 +162,99 @@ func (b *Boolean) Equal(value bool) *Boolean {
+	return b
+}
+
+func boolList(values []bool) []interface{} {
+	l := make([]interface{}, 0, len(values))
+	for _, v := range values {
+		l = append(l, v)
+	}
+
+	return l
+}
```
Nitpick: please move this to the end of file.

fahimbagar replied:
addressed in https://github.com/gavv/httpexpect/pull/256/commits/0d5cba174efafb67610d8f5eda37b1fa4ea8e193

gavv commented on array.go:
```diff
@@ -746,6 +746,128 @@ func (a *Array) Equal(value interface{}) *Array {
+			return a
+		}
+
+		if reflect.DeepEqual(expected, a.value) {
+			isListed = true
+		}
```
Let's add `break` here and in similar places

gavv commented on boolean.go:
```diff
@@ -162,6 +162,99 @@ func (b *Boolean) Equal(value bool) *Boolean {
+		return b
+	}
+
+	for _, v := range values {
+		if b.value == v {
+			return b
+		}
+	}
+
+	opChain.fail(AssertionFailure{
+		Type:     AssertBelongs,
+		Actual:   &AssertionValue{b.value},
+		Expected: &AssertionValue{AssertionList(boolList(values))},
+		Errors: []error{
+			errors.New("expected: boolean is listed"),
+		},
+	})
+
+	return b
```
Nitpick: usually httpexpect methods use the following pattern:

* method can have early returns in case of failures
* there is only one successful return, at the end of the method

You used exactly this layout in Array.InList, but in Boolean.InList you used the opposite layout: early return for success, failure in the end of the method.

I suggest to refactor all InList to follow the same layout as in Array.InList, for consistency.

fahimbagar replied:
Thanks for the review.

Previously, I tried to use early returns but I had a doubt since types like `Array` receives `interface{}` as param, so I tried to iterate everything first to check each of the params and then decided whether it's succeeded.

I will take this as a guideline
> - method can have early returns in case of failures
> - there is only one successful return, at the end of the method

fahimbagar replied:
`InList` uses `isListed` and breaks early in each type. While `NotInList` add failed step and breaks early if same.

Let me know if I miss something there, @gavv.

gavv reviewed (commented):
Thanks for update! A few comments

gavv commented:
Thanks!

gavv commented:
Small follow-up commit: 8fe2ffba3601d60a56a162077d20f4a40833233b

gavv commented:
PS. If you like chat format, welcome to our [discord](https://discord.gg/5SCPCuCWA9) :)
